File: bot-api/dotnet/test/src/util/JsonConverterTest.cs
using System;
using NUnit.Framework;
using Robocode.TankRoyale.BotApi.Internal.Json;
using Robocode.TankRoyale.BotApi.Util;
using Robocode.TankRoyale.Schema;

namespace Robocode.TankRoyale.BotApi.Tests.Util;

[Description("TR-API-UTL-002 JsonUtil serialization + schema compliance")]
public class JsonConverterTest
{
    [Test]
    [Category("UTL")] 
    [Property("ID", "TR-API-UTL-002")]
    public void GivenScannedBotEvent_whenSerializedAndDeserialized_thenFieldsArePreserved()
    {
        // Arrange
        var evt = new ScannedBotEvent
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.ScannedBotEvent),
            TurnNumber = 1,
            ScannedByBotId = 2,
            ScannedBotId = 3,
            Energy = 100,
            X = 10,
            Y = 20,
            Direction = 90,
            Speed = 5,
        };

        // Act
        var json = JsonConverter.ToJson(evt);
        var deserialized = JsonConverter.FromJson(json, typeof(ScannedBotEvent)) as ScannedBotEvent;

        // Assert
        Assert.That(deserialized, Is.Not.Null);
        Assert.Multiple(() =>
        {
            Assert.That(deserialized.Type, Is.EqualTo(evt.Type));
            Assert.That(deserialized.TurnNumber, Is.EqualTo(evt.TurnNumber));
            Assert.That(deserialized.ScannedByBotId, Is.EqualTo(evt.ScannedByBotId));
            Assert.That(deserialized.ScannedBotId, Is.EqualTo(evt.ScannedBotId));
            Assert.That(deserialized.Energy, Is.EqualTo(evt.Energy));
            Assert.That(deserialized.X, Is.EqualTo(evt.X));
            Assert.That(deserialized.Y, Is.EqualTo(evt.Y));
            Assert.That(deserialized.Direction, Is.EqualTo(evt.Direction));
            Assert.That(deserialized.Speed, Is.EqualTo(evt.Speed));
        });
    }

    [Test]
    [Category("UTL")] 
    [Property("ID", "TR-API-UTL-002")]
    public void GivenBulletFiredEventWithNestedBulletState_whenSerializedAndDeserialized_thenNestedFieldsArePreserved()
    {
        // Arrange
        var bullet = new global::Robocode.TankRoyale.Schema.BulletState
        {
            BulletId = 1,
            OwnerId = 2,
            Power = 2.5,
            X = 100,
            Y = 200,
            Direction = 45,
            Color = "#FF0000",
        };
        var evt = new BulletFiredEvent
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.BulletFiredEvent),
            TurnNumber = 7,
            Bullet = bullet,
        };

        // Act
        var json = JsonConverter.ToJson(evt);
        var deserialized = JsonConverter.FromJson(json, typeof(BulletFiredEvent)) as BulletFiredEvent;

        // Assert
        Assert.That(deserialized, Is.Not.Null);
        Assert.That(deserialized.Bullet, Is.Not.Null);
        Assert.Multiple(() =>
        {
            Assert.That(deserialized.Type, Is.EqualTo(evt.Type));
            Assert.That(deserialized.TurnNumber, Is.EqualTo(evt.TurnNumber));
            Assert.That(deserialized.Bullet.BulletId, Is.EqualTo(bullet.BulletId));
            Assert.That(deserialized.Bullet.OwnerId, Is.EqualTo(bullet.OwnerId));
            Assert.That(deserialized.Bullet.Power, Is.EqualTo(bullet.Power));
            Assert.That(deserialized.Bullet.X, Is.EqualTo(bullet.X));
            Assert.That(deserialized.Bullet.Y, Is.EqualTo(bullet.Y));
            Assert.That(deserialized.Bullet.Direction, Is.EqualTo(bullet.Direction));
            Assert.That(deserialized.Bullet.Color, Is.EqualTo(bullet.Color));
        });
    }
}

File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/internal/json/JsonConverterTest.java
package dev.robocode.tankroyale.botapi.internal.json;

import dev.robocode.tankroyale.schema.*;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayName("TR-API-UTL-002 JsonUtil serialization + schema compliance")
class JsonConverterTest {

    @Test
    @Tag("UTL")
    @Tag("TR-API-UTL-002")
    void givenScannedBotEvent_whenSerializedAndDeserialized_thenFieldsArePreserved() {
        // Arrange
        var evt = new ScannedBotEvent();
        evt.setType(Message.Type.SCANNED_BOT_EVENT);
        evt.setTurnNumber(1);
        evt.setScannedByBotId(2);
        evt.setScannedBotId(3);
        evt.setEnergy(100.0);
        evt.setX(10.0);
        evt.setY(20.0);
        evt.setDirection(90.0);
        evt.setSpeed(5.0);

        // Act
        var json = JsonConverter.toJson(evt);
        var deserialized = JsonConverter.fromJson(json, ScannedBotEvent.class);

        // Assert
        assertThat(deserialized).isNotNull();
        assertThat(deserialized.getType()).isEqualTo(evt.getType());
        assertThat(deserialized.getTurnNumber()).isEqualTo(evt.getTurnNumber());
        assertThat(deserialized.getScannedByBotId()).isEqualTo(evt.getScannedByBotId());
        assertThat(deserialized.getScannedBotId()).isEqualTo(evt.getScannedBotId());
        assertThat(deserialized.getEnergy()).isEqualTo(evt.getEnergy());
        assertThat(deserialized.getX()).isEqualTo(evt.getX());
        assertThat(deserialized.getY()).isEqualTo(evt.getY());
        assertThat(deserialized.getDirection()).isEqualTo(evt.getDirection());
        assertThat(deserialized.getSpeed()).isEqualTo(evt.getSpeed());
    }

    @Test
    @Tag("UTL")
    @Tag("TR-API-UTL-002")
    void givenBulletFiredEventWithNestedBulletState_whenSerializedAndDeserialized_thenNestedFieldsArePreserved() {
        // Arrange
        var bullet = new BulletState();
        bullet.setBulletId(1);
        bullet.setOwnerId(2);
        bullet.setPower(2.5);
        bullet.setX(100.0);
        bullet.setY(200.0);
        bullet.setDirection(45.0);
        bullet.setColor("#FF0000");

        var evt = new BulletFiredEvent();
        evt.setType(Message.Type.BULLET_FIRED_EVENT);
        evt.setTurnNumber(7);
        evt.setBullet(bullet);

        // Act
        var json = JsonConverter.toJson(evt);
        var deserialized = JsonConverter.fromJson(json, BulletFiredEvent.class);

        // Assert
        assertThat(deserialized).isNotNull();
        assertThat(deserialized.getType()).isEqualTo(evt.getType());
        assertThat(deserialized.getTurnNumber()).isEqualTo(evt.getTurnNumber());
        assertThat(deserialized.getBullet()).isNotNull();
        assertThat(deserialized.getBullet().getBulletId()).isEqualTo(bullet.getBulletId());
        assertThat(deserialized.getBullet().getOwnerId()).isEqualTo(bullet.getOwnerId());
        assertThat(deserialized.getBullet().getPower()).isEqualTo(bullet.getPower());
        assertThat(deserialized.getBullet().getX()).isEqualTo(bullet.getX());
        assertThat(deserialized.getBullet().getY()).isEqualTo(bullet.getY());
        assertThat(deserialized.getBullet().getDirection()).isEqualTo(bullet.getDirection());
        assertThat(deserialized.getBullet().getColor()).isEqualTo(bullet.getColor());
    }
}
